exonum-java-binding/core/rust/src/handle/resource_manager/imp.rs

```diff
@@ -94,9 +99,12 @@ fn check_handle_impl<T: 'static>(handle: Handle, ownership: Option<HandleOwnersh
         Some(info) => {
             let actual_object_type = TypeId::of::<T>();
             assert_eq!(
-                info.object_type, actual_object_type,
-                "Wrong type id for '{:X}' handle",
-                handle
+                info.object_type,
+                actual_object_type,
+                "Wrong type id for '{:X}' handle, expected '{}', actual '{}'",
+                handle,
+                info.type_name,
+                any::type_name::<T>(),
```

Comment:
https://doc.rust-lang.org/std/any/fn.type_name.html — is it 1.38.0 new feature (is that grey box a 'since' field)? We are stuck on 1.36.0 at the moment.

Reply:
Yes
```
#[stable(feature = "type_name", since = "1.38.0")]
#[rustc_const_unstable(feature = "const_type_name")]
pub const fn type_name<T: ?Sized>() -> &'static str {
```

Reply:
Yes, it is new feature unfortunately.

Reply:
OK, thanks, then we will have to wait till a newer working compiler is available 🙃

Reply:
\+ @vitvakatu 

It seems it is no longer blocked since #1429.

> A note to the one who will be syncing ~~dynamic-services~~ this PR with master: dynamic-services got rid of handles on native-owned objects (#1178), and of HandleInfo as well, hence this file will have to be merged manually.

I also wonder if it will make sense to keep string representations separately to avoid their duplication in handles of the same type.

Reply:
I can easily miss something, but isn't "resource manager" still a debug tool? I guess such optimization isn't necessary in that case.

Though it can be interesting to measure overhead in the typical use-case.

Reply:
> I can easily miss something, but isn't "resource manager" still a debug tool? I guess such optimization isn't necessary in that case.

Yes, it is primarily a debug tool (actually, a useful one — we recently had an issue with invalid handles during development 🙃). However, there are currently quite some publicly accessible APIs that allow to pass anything as handle to native from the user code. A normal user has absolutely no reason to use these API and likely never won't, but someone _might_. On the other hand, I don't know if that shall make us keep the rm in production builds 🤔